Sources/Testing/Test.swift

```diff
@@ -68,24 +68,91 @@ public struct Test: Sendable {
   @_spi(ForToolsIntegrationOnly)
   public var xcTestCompatibleSelector: __XCTestCompatibleSelector?
 
-  /// Storage for the ``testCases`` property.
+  /// An enumeration describing the evaluation state of a test's cases.
   ///
-  /// This use of `UncheckedSendable` and of `AnySequence` is necessary because
-  /// it is not currently possible to express `Sequence<Test.Case> & Sendable`
-  /// as an existential (`any`) ([96960993](rdar://96960993)). It is also not
-  /// possible to have a value of an underlying generic sequence type without
-  /// specifying its generic parameters.
-  private var _testCases: UncheckedSendable<AnySequence<Test.Case>>?
+  /// This use of `UncheckedSendable` and of `AnySequence` in this type's cases
+  /// is necessary because it is not currently possible to express
+  /// `Sequence<Test.Case> & Sendable` as an existential (`any`)
+  /// ([96960993](rdar://96960993)). It is also not possible to have a value of
+  /// an underlying generic sequence type without specifying its generic
+  /// parameters.
+  fileprivate enum TestCasesState: Sendable {
+    /// The test's cases have not yet been evaluated.
+    ///
+    /// - Parameters:
+    ///   - function: The function to call to evaluate the test's cases. The
+    ///     result is a sequence of test cases.
+    case unevaluated(_ function: @Sendable () async throws -> AnySequence<Test.Case>)
+
+    /// The test's cases have been evaluated, and either returned a sequence of
+    /// cases or failed by throwing an error.
+    ///
+    /// - Parameters:
+    ///   - result: The result of having evaluated the test's cases.
+    case evaluated(_ result: Result<UncheckedSendable<AnySequence<Test.Case>>, any Error>)
+  }
+
+  /// The evaluation state of this test's cases, if any.
+  ///
+  /// If this test represents a suite type, the value of this property is `nil`.
+  fileprivate var testCasesState: TestCasesState?
 
   /// The set of test cases associated with this test, if any.
   ///
+  /// - Precondition: This property may only be accessed on test instances
+  ///   representing suites, or on test functions whose ``testCaseState``
+  ///   indicates a successfully-evaluated state.
+  ///
   /// For parameterized tests, each test case is associated with a single
   /// combination of parameterized inputs. For non-parameterized tests, a single
   /// test case is synthesized. For test suite types (as opposed to test
   /// functions), the value of this property is `nil`.
-  @_spi(ForToolsIntegrationOnly)
-  public var testCases: (some Sequence<Test.Case> & Sendable)? {
-    _testCases?.rawValue
+  var testCases: (some Sequence<Test.Case> & Sendable)? {
+    guard let testCasesState else {
+      return nil as AnySequence<Test.Case>?
```

Comment:
Is the `as` clause necessary here?

Reply:
Compiler can't infer the `some` result type without it, since this is the first `return` statement. (This technique was actually needed in the previous commit in this PR too, just in a different place)

Reply:
Boooo okay.